tools/wake/cli_options.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <vector>
+
+#include "gopt/gopt-arg.h"
+#include "gopt/gopt.h"
+
+struct CommandLineOptions {
+  bool check;
+  bool verbose;
+  bool debug;
+  bool quiet;
+  bool wait;
+  bool workspace;
+  bool tty;
+  bool fwarning;
+  int profileh;
+  unsigned int input;
+  unsigned int output;
+  bool last_use;
+  bool last_exe;
+  bool lsp;
+  bool failed;
+  bool script;
+  bool metadata;
+  bool version;
+  bool html;
+  bool global;
+  bool help;
+  bool config;
+  bool debugdb;
+  bool parse;
+  bool tcheck;
+  bool dumpssa;
+  bool optim;
+  bool exports;
+  bool timeline;
+  bool clean;
+  bool list_outputs;
+
+  const char *percent_str;
+  const char *jobs_str;
+  const char *memory_str;
+  const char *heapf;
+  const char *profile;
+  const char *init;
+  const char *chdir;
+  const char *in;
+  const char *exec;
+  const char *job;
+  const char *label;
+  char *shebang;
+  const char *tagdag;
+  const char *tag;
+  const char *api;
+  const char *fd1;
+  const char *fd2;
+  const char *fd3;
+  const char *fd4;
+  const char *fd5;
+
+  std::vector<char *> input_files;
+  std::vector<char *> output_files;
+
+  int argc;
+  char **argv;
+
+  CommandLineOptions(int argc_in, char **argv_in) {
+    argv = argv_in;
+    unsigned int max_pairs = argc_in / 2;
```

Comment:
This works for a kind of subtle reason. Because argc_in is actully 1 higher than the logical length of argv_in because the 0th argument is the tool and not the arguments you don't have to round up. In general though, calling `reserve` is not a good idea. It's generally a que to me taht someone is about to do something very wrong with a vector or someone is over-optimizing something that they shouldn't be. I would recommend removing these 3 lines.